qlib/kernel/threadmgr/task_signals.rs

```diff
@@ -1263,9 +1263,115 @@ impl Task {
         return Err(Error::SysCallRetCtrl(TaskRunState::RunSyscallRet));
     }
 
+
+    #[cfg(target_arch="aarch64")]
+    pub fn deliverSignalToHandler(&mut self, info: &SignalInfo, sigAct: &SigAct) -> Result<()> {
+        debug!(">>> deliver signal: no {}\n sigAct {:#?}", info.Signo, sigAct);
+        // set up signal frame
+        let pt = self.GetPtRegs();
+        let mut userStack = Stack::New(pt.get_stack_pointer() - 128); // red zone
+
+        // NOTE sigaltstack
+        if sigAct.flags.IsOnStack() && self.signalStack.IsEnable() {
+            self.signalStack.SetOnStack();
+            if !self.signalStack.Contains(pt.get_stack_pointer()) {
+                userStack = Stack::New(self.signalStack.Top());
+            }
+        }
+
+
+        debug!(">>> deliver signal: user stack {:#X}, SignalStack {:#?}", userStack.sp, self.signalStack);
+        // let's forget about the fp units for the moment
+
+        let t = self.Thread();
+        let mut mask = t.lock().signalMask;
+        let haveSavedSignalMask = t.lock().haveSavedSignalMask;
+        if haveSavedSignalMask {
+            mask = t.lock().savedSignalMask;
+            t.lock().haveSavedSignalMask = false;
+        }
+
+        let mut newMask = t.lock().signalMask;
+        newMask.0 |= sigAct.mask;
+        if !sigAct.flags.IsNoDefer() {
+            newMask.0 |= SignalSet::New(Signal(info.Signo)).0;
+        }
+        t.SetSignalMask(newMask);
+
+        let mut fault_addr = 0;
+        if info.Signo == Signal::SIGBUS || info.Signo == Signal::SIGSEGV {
+            fault_addr = info.SigFault().addr;
+        }
+
+        // save old user state
+        // PtRegs, oldMask, cr2, fpstate, alt.
+        debug!(">>> deliver signal: before UContext init");
+        let mut ctx = UContext::New(pt, mask.0, fault_addr, 0, &self.signalStack);
+        debug!(">>> deliver signal: stored UContext: {:#X?}", ctx);
+        let sigInfoAddr = userStack.PushType::<SignalInfo>(self, info)?;
+        let sigCtxAddr = userStack.PushType::<UContext>(self, &ctx)?;
+
+        let signo = info.Signo as u64;
+        // let sp = userStack.PushU64(self, sigAct.restorer)?;
+        // jump to user handler
+        let currTask = Task::Current();
+        let regs = currTask.GetPtRegs();
+        *regs = PtRegs::default();
+        // TODO maybe need to handle PSTATE?
+        // regs.sp = sp;
+        regs.sp = sigCtxAddr;
+        regs.pc = sigAct.handler;
+        // TODO set proper value for PSTATE
+        regs.pstate = 0x0;
+        // parameters to the handler
+        // handler(int signo, siginfo_t *info, void *context)
+        debug!(">>> deliver signal: restorer is {:#X}", sigAct.restorer);
+        regs.regs[0] = signo;
+        regs.regs[1] = sigInfoAddr;
+        regs.regs[2] = sigCtxAddr;
+        regs.regs[30] = sigAct.restorer;
```

Comment:
If I understand _[sygaction(2)](https://man7.org/linux/man-pages/man2/sigaction.2.html)_, for x86_64 the _.restore_ field is valid if the _SA_RESTORE_ flag is set. For aarch64, that uses the vDSO, should it be taken as default?

Reply:
it's not the kernel's responsibility to set the restorer addr to vdso function  here. The user library parses vdso, and give the address to sigaction.

Reply:
Then I don't know. This was the most obvious arch. difference in the manual.

Reply:
> it's not the kernel's responsibility to set the restorer addr to vdso function here. The user library parses vdso, and give the address to sigaction.

I stand corrected. the kernel should point the trampoline to vdso symbols if restorer is not present. Will fix.

Reply:
I believe _xmas-elf_ crate can be used to read the symbol table of the vDSO, and get the symbol. Ping me if I should provide a patch.

Reply:
I have a temporary workaround only for the signals. 

I don't want to add extensive VDSO support in this series. Could you perhaps implement a module that returns address from loaded user vdso by symbol name (str) ?  Please have another PR if you like.

Something like linux kernel `arch/arm64/include/asm/vdso.h` would be nice. (Perhaps store the VDSO base in task.mm)

```
#include <generated/vdso-offsets.h>
#define VDSO_SYMBOL(base, name)						   \
({									   \
	(void *)(vdso_offset_##name - VDSO_LBASE + (unsigned long)(base)); \
})
```

Also take a look at 

```
pub fn GetVDSOParamPageAddr() -> u64 {
    return TIME_KEEPER.read().params.GetParamPageAddr();
}
```

from `qlib/kernel/kernel/timer/mod.rs`. Since VDSO is not only about time, I'd like to have relevant code/data in in a different place.

Reply:
I have an additional field in task.mm called `UserVDSOBase`. This value is loaded when a user task is loaded.

Reply:
I 'll look into it.